src/iterator.jl

```diff
@@ -0,0 +1,38 @@
+struct CuIterator{B}
+    batches::B
+    initial_pool_size::Ref{Int}
+    pool::Ref{Mem.DeviceBuffer}
+end
+
+function CuIterator(batches, initial_pool_size::Integer=0)
+    return CuIterator(batches, Ref(initial_pool_size), Ref{Mem.DeviceBuffer}())
+end
+
+function Base.iterate(c::CuIterator, state...)
+    item = iterate(c.batches, state...)
+    if item === nothing
+        Mem.free(c.pool[])
+        return nothing
+    end
+    batch, next_state = item
+    required_pool_size = sum(sizeof, batch)
+    if isempty(state)
+        c.initial_pool_size[] = max(required_pool_size, c.initial_pool_size[])
+        c.pool[] = Mem.alloc(Mem.DeviceBuffer, c.initial_pool_size[])
+    elseif required_pool_size > sizeof(c.pool[])
+        Mem.free(c.pool[])
+        c.initial_pool_size[] = required_pool_size
+        c.pool[] = Mem.alloc(Mem.DeviceBuffer, required_pool_size)
```

Comment:
Does not free the old pool?

Reply:
Ah, I thought the `Mem.free(c.pool[])` above would?

Reply:
Oh oops, I must be blind.